packages/flutter/lib/src/widgets/pop_scope.dart

```diff
@@ -0,0 +1,162 @@
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [NavigatorPopHandler], which is a less verbose way to handle system back
+///    gestures in the case of nested [Navigator]s.
+///  * [Form.canPop] and [Form.onPopInvoked], which can be used to handle system
+///    back gestures in the case of a form with unsaved data.
+///  * [ModalRoute.registerPopInterface] and [ModalRoute.unregisterPopInterface],
+///    which this widget uses to integrate with Flutter's navigation system.
+class PopScope extends StatefulWidget {
+  /// Creates a widget that registers a callback to veto attempts by the user to
+  /// dismiss the enclosing [ModalRoute].
+  ///
+  /// The [child] argument must not be null.
+  const PopScope({
+    super.key,
+    required this.child,
+    this.canPop = true,
+    this.onPopInvoked,
+  });
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// {@template flutter.widgets.PopScope.onPopInvoked}
+  /// Called after a route pop was handled.
+  /// {@endtemplate}
+  ///
+  /// It's not possible to prevent the pop from happening at the time that this
+  /// method is called; the pop has already happened. Use [canPop] to
+  /// disable pops in advance.
+  ///
+  /// This will still be called even when the pop is canceled. A pop is canceled
+  /// when the relevant [Route.popDisposition] returns false, such as when
+  /// [canPop] is set to false on a [PopScope]. The `didPop` parameter
+  /// indicates whether or not the back navigation actually happened
+  /// successfully.
+  ///
+  /// See also:
+  ///
+  ///  * [Route.onPopInvoked], which is similar.
+  final PopInvokedCallback? onPopInvoked;
+
+  /// {@template flutter.widgets.PopScope.canPop}
+  /// When false, blocks the current route from being popped.
+  ///
+  /// This includes the root route, where upon popping, the Flutter app would
+  /// exit.
+  ///
+  /// If multiple [PopScope] widgets appear in a route's widget subtree, then
+  /// each and every `canPop` must be `true` in order for the route to be
+  /// able to pop.
+  ///
+  /// [Android's predictive back](https://developer.android.com/guide/navigation/predictive-back-gesture)
+  /// feature will not animate when this boolean is false.
+  /// {@endtemplate}
+  final bool canPop;
+
+  @override
+  State<PopScope> createState() => _PopScopeState();
+}
+
+class _PopScopeState extends State<PopScope> implements PopInterface {
+  ModalRoute<dynamic>? _route;
+
+  @override
+  PopInvokedCallback? get onPopInvoked => widget.onPopInvoked;
+
+  @override
+  late final ValueNotifier<bool> canPopNotifier;
+
+  @override
+  void initState() {
+    super.initState();
+    canPopNotifier = ValueNotifier<bool>(widget.canPop);
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _route?.unregisterPopInterface(this);
+    _route = ModalRoute.of(context);
+    _route?.registerPopInterface(this);
```

Comment:
just curious, have you thought about reuse the localhistory entry logic? or consolidate both APIs.

Reply:
I wasn't familiar with it, I just went with this approach because it was the same as WillPopScope.  It looks pretty close to what I want here though...  Playing around with it I think I see two potential problems:

  1. I need to add/remove based on the presence of PopScope in the widget tree, but if I try to call addLocalHistoryEntry during initState, I get the `setState called during build` error.
  2. There is some built-in machinery for removing the last LocalHistoryEntry on pop that is kind of fundamentally different than how I'd be using this.  Likewise, PopScope's need to be called whenever a back gesture happens is something that would be new to LocalHistoryEntry.

Given that this is just swapping WillPopScope for PopScope I'm not inclined to reconcile these differences now, though there probably is some common abstraction that would work for both of them.